From: Rafael J. Wysocki <rafael.j.wysocki@intel.com>
Subject: kexec_core: fix error code path in the KEXEC_JUMP flow
Date: Thu, 10 Jul 2025 15:10:41 +0200

If dpm_suspend_start() fails, dpm_resume_end() must be called to recover
devices whose suspend callbacks have been called, but this does not happen
in the KEXEC_JUMP flow's error path due to a confused goto target label.

Address this by using the correct target label in the goto statement in
question.

Link: https://lkml.kernel.org/r/2396879.ElGaqSPkdT@rjwysocki.net
Fixes: 2965faa5e03d ("kexec: split kexec_load syscall from kexec core code")
Signed-off-by: Rafael J. Wysocki <rafael.j.wysocki@intel.com>
Cc: Dave Young <dyoung@redhat.com>
Cc: Baoquan He <bhe@redhat.com>
Cc: David Woodhouse <dwmw2@infradead.org>
Cc: Mario Limonciello <mario.limonciello@amd.com>
Cc: Stephen Rothwell <sfr@canb.auug.org.au>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 kernel/kexec_core.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/kernel/kexec_core.c~kexec_core-fix-error-code-path-in-the-kexec_jump-flow
+++ a/kernel/kexec_core.c
@@ -1080,7 +1080,7 @@ int kernel_kexec(void)
 		console_suspend_all();
 		error = dpm_suspend_start(PMSG_FREEZE);
 		if (error)
-			goto Resume_console;
+			goto Resume_devices;
 		/*
 		 * dpm_suspend_end() must be called after dpm_suspend_start()
 		 * to complete the transition, like in the hibernation flows
_
